java/dataset/src/main/java/org/apache/arrow/dataset/jni/DirectReservationListener.java

```diff
@@ -87,11 +87,47 @@ public void unreserve(long size) {
   public long getCurrentDirectMemReservation() {
     try {
       final Class<?> classBits = Class.forName("java.nio.Bits");
-      final Field f = classBits.getDeclaredField("reservedMemory");
+      final Field f = this.getDeclaredFieldBaseOnJDKVersion(classBits, "reservedMemory");
       f.setAccessible(true);
       return ((AtomicLong) f.get(null)).get();
     } catch (Exception e) {
       throw new RuntimeException(e);
     }
   }
+
+  /**
+   * To evaluate method on a class base on JDK version.
+   * @param classBits Object associated with the class with the given string name
+   * @param name Method needed to evaluate
+   * @return
+   */
+  private Method getDeclaredMethodBaseOnJDKVersion(Class<?> classBits, String name) {
+    try {
+      return classBits.getDeclaredMethod(name, long.class, int.class);
+    } catch (NoSuchMethodException e) {
+      try {
+        return classBits.getDeclaredMethod(name, long.class, long.class);
+      } catch (NoSuchMethodException ex) {
+        throw new AssertionError(ex);
+      }
+    }
+  }
+
+  /**
+   * To evaluate field on a class base on JDK version.
+   * @param classBits Object associated with the class with the given string name
+   * @param name Field needed to evaluate
+   * @return
+   */
+  private Field getDeclaredFieldBaseOnJDKVersion(Class<?> classBits, String name) {
```

Comment:
This isn't a general-purpose method, so I would just inline it, or at least name it so it's clear it's specifically for the reservedMemory field. (And remove the second parameter in that case.)

Reply:
Deleted